### DISCUSSION TITLE
Using SPI Bridge with HyperSerialWLED

### DISCUSSION THREAD

**Discussion by peetipablo** [awawa-dev/HyperHDR#219] [Q&A]
Hello!  I have a working setup with HyperHDR, and I have a question for you if you don't mind!  My setup is below:

- Raspberry Pi 4
- SK6812 LED strip (266 LED)
- EZcap 320 B USB Grabber
- esp8266 D1 mini clone flashed with HyperSPI

I was not able to get this esp8266 working with HyperSerialEsp8266.  I was able to get this esp8266 working with HyperSerialWLED (WLED protocol ONLY, not adalight protocol).  Finally wiring the pi/esp8266 for SPI and flashing with HyperSPI worked.  Reading through some of the discussions and issues in your github repos I believe that I have counterfeit esp8266 from amazon that does not support 2M bandwidth.

Is there any way to continue to use the SPI connection (non-USB), but also HyperSerialWLED flashed on the esp8266?  I could not find anything related to this in your documentation, but if I missed something I apologize.

Thank you for all the great work you've done on these projects, it is truly amazing!

**Comment by awawa-dev** [comment]
Hi
You may be right about the reason why HyperSerialEsp8266 didn't work. In HyperSerialWLED you can find special release using slower speed (special_version_speed_921600_only_HyperSerialWLED_ESP8266) than should work on 921600 instead of 2Mb: that speed should be also set in HyperHDR Adalight driver for that release. Usually that version works fine on some problematic ESP boards. You can also modify and compile HyperSerialEsp8266 for 921600 if it works.

> Is there any way to continue to use the SPI connection (non-USB), but also HyperSerialWLED flashed on the esp8266? I could not find anything related to this in your documentation, but if I missed something I apologize.

One ESP device could not handle both SPI and adalight serial communication so 2 ESP devices are needed (one for SPI and the other for HyperSerialEsp8266 or HyperSerialEsp32). While I haven't tested the exact same setup, I don't see any reason why it shouldn't work: just create 2 instances in HyperHDR.

**Comment by peetipablo** [comment]
Thanks for the quick response!  Can I also just pick your brain to see what
the real world performance impact of using the 921600 speed vs 2000000
speed serial driver would be? Will I see any major performance difference
using hyperSPI vs HyperSerialEsp8266?

**Reply by awawa-dev** [reply]
HyperSPI allows to utilize full bandwidth of sk6812 thanks to much higher speed, less latency and less protocol overheat then HyperSerialEsp over USB.  Theoretically even HyperSerial 2Mb communication speed should fully saturate sk6812 bus later, but it does not happen: HyperSPI can provide 5-10% higher refresh rate. 921600  is still faster than internal sk6812 ~800Kb (and WLED with 115200)  but didn't test its performance because I treat it as fallback solution when everything else fails. You can read you current refresh rate connecting to HyperSPI using USB serial port and compare it later to WLED@921600  FPS available in the info panel.
